src/plugins/discover/public/utils/get_fields_to_show.ts

```diff
@@ -8,24 +8,38 @@
 import { getFieldSubtypeMulti } from '@kbn/data-views-plugin/public';
 import type { DataView } from '@kbn/data-views-plugin/public';
 
+/**
+ * Returns am array of fields to display in the Documents column of the data table
+ * If showMultiFields is set to false, it filters out multifields that have a parent, to prevent entries for multifields
+ * like this: field, field.keyword, field.whatever
+ * @param fields
+ * @param dataView
+ * @param showMultiFields
+ */
 export const getFieldsToShow = (fields: string[], dataView: DataView, showMultiFields: boolean) => {
-  const childParentFieldsMap = {} as Record<string, string>;
-  const mapping = (name: string) => dataView.fields.getByName(name);
+  if (showMultiFields) {
+    return fields;
+  }
+  const fieldSet = new Set();
+  const childParentFieldsMap = new Map();
+  const parentFieldSet = new Set();
   fields.forEach((key) => {
-    const mapped = mapping(key);
+    const mapped = dataView.fields.getByName(key);
     const subTypeMulti = mapped && getFieldSubtypeMulti(mapped.spec);
+    const isMultiField = Boolean(subTypeMulti?.multi);
     if (mapped && subTypeMulti?.multi?.parent) {
-      childParentFieldsMap[mapped.name] = subTypeMulti.multi.parent;
+      childParentFieldsMap.set(key, subTypeMulti.multi.parent);
+    }
+    if (mapped && isMultiField) {
+      parentFieldSet.add(key);
     }
+    fieldSet.add(key);
   });
   return fields.filter((key: string) => {
-    const fieldMapping = mapping(key);
-    const subTypeMulti = fieldMapping && getFieldSubtypeMulti(fieldMapping.spec);
-    const isMultiField = !!subTypeMulti?.multi;
-    if (!isMultiField) {
+    if (!parentFieldSet.has(key)) {
       return true;
     }
-    const parent = childParentFieldsMap[key];
-    return showMultiFields || (parent && !fields.includes(parent));
+    const parent = childParentFieldsMap.get(key);
+    return parent && !fieldSet.has(parent);
```

Comment:
We could simplify even more as the following:
```
export const getFieldsToShow = (fields: string[], dataView: DataView, showMultiFields: boolean) => {
  if (showMultiFields) {
    return fields;
  }
  return fields.filter((fieldName) => {
    const mapped = dataView.fields.getByName(fieldName);
    const isMultiField = Boolean(mapped && getFieldSubtypeMulti(mapped.spec));
    return !isMultiField;
  });
};
```

It's the same way as we filter subfields from the sidebar.

Reply:
This is a great suggestion, it further simplifies the code, removes another iteration of the field list ... and works, applied the change

Reply:
@majagrubic this change seems to break a legacy table test 

https://github.com/elastic/kibana/blob/bcbef78b4dfaebe40f334fcaa5cef94fda9acc29/src/plugins/discover/public/services/doc_views/components/doc_viewer_table/legacy/table.test.tsx#L469

It expects `city.raw` to be rendered, because `city` is not returned by the fields API .. but `city` is in the mapping. I simply can't remember why we were doing this and if this is an actual case to be considered? 

@jughosta To be safe we could revert the last change for this PR, apply it in a follow up for the next minor, without backporting ... getting rid of legacy

Reply:
Sounds good!

Reply:
@jughosta wanna create the follow up PR with your suggestion, just for 8.7?

Reply:
Follow up PR https://github.com/elastic/kibana/pull/145698
I kept the parent-checking logic too.